include/flang/lower/ASTBuilder.h

```diff
@@ -0,0 +1,332 @@
+      : u{&a}, parent{p}, pos{pos}, lab{lab} {}
+
+  /// Compiler-generated jump
+  Evaluation(const CGJump &jump, const ParentType &p)
+      : u{jump}, parent{p}, cfg{CFGAnnotation::Goto} {}
+
+  /// Construct ctor
+  template <typename A>
+  Evaluation(const A &a, const ParentType &parent) : u{&a}, parent{parent} {
+    static_assert(AST::isConstruct<A>(), "must be a construct");
+  }
+
+  /// is `A` executable (an action statement or compiler generated)?
+  template <typename A>
+  constexpr static bool isAction(const A &a) {
+    return !AST::isConstruct<A>() && !isOther(a);
+  }
+
+  /// is `A` a compiler-generated evaluation?
+  template <typename A>
+  constexpr static bool isGenerated(const A &) {
+    return std::is_same_v<A, CGJump>;
+  }
+
+  /// is `A` not an executable statement?
+  template <typename A>
+  constexpr static bool isOther(const A &) {
+    return std::is_same_v<A, parser::FormatStmt> ||
+           std::is_same_v<A, parser::EntryStmt> ||
+           std::is_same_v<A, parser::DataStmt> ||
+           std::is_same_v<A, parser::NamelistStmt>;
+  }
+
+  constexpr bool isActionStmt() const {
+    return std::visit(common::visitors{
+                          [](auto *p) { return isAction(*p); },
+                          [](auto &r) { return isGenerated(r); },
+                      },
+                      u);
+  }
+
+  constexpr bool isStmt() const {
+    return std::visit(common::visitors{
+                          [](auto *p) { return isAction(*p) || isOther(*p); },
+                          [](auto &r) { return isGenerated(r); },
+                      },
+                      u);
+  }
+  constexpr bool isConstruct() const { return !isStmt(); }
+
+  /// Set the type of originating control flow type for this evaluation.
+  void setCFG(CFGAnnotation a, Evaluation *cstr) {
+    cfg = a;
+    setBranches(cstr);
+  }
+
+  /// Is this evaluation a control-flow origin? (The AST must be annotated)
+  bool isControlOrigin() const { return cfg != CFGAnnotation::None; }
+
+  /// Is this evaluation a control-flow target? (The AST must be annotated)
+  bool isControlTarget() const { return isTarget; }
+
+  /// Set the containsBranches flag iff this evaluation (a construct) contains
+  /// control flow
+  void setBranches() { containsBranches = true; }
+
+  constexpr std::list<Evaluation> *getConstructEvals() {
+    return isStmt() ? nullptr : subs;
+  }
+
+  /// Set that the construct `cstr` (if not a nullptr) has branches.
+  static void setBranches(Evaluation *cstr) {
+    if (cstr) {
+      cstr->setBranches();
+    }
+  }
+
+  EvalVariant u;
+  ParentType parent;
+  parser::CharBlock pos;
+  std::optional<parser::Label> lab;
+  std::list<Evaluation> *subs{nullptr}; // construct sub-statements
+  CFGAnnotation cfg{CFGAnnotation::None};
+  bool isTarget{false};         // this evaluation is a control target
+  bool containsBranches{false}; // construct contains branches
+};
+
+/// A program is a list of program units.
```

Comment:
I have explored your suggestion a bit, I see the following differences:
1. The "AST" unit class have a pointer to the "end" nodes, the `ProgramTree` does not.
2. One of the "AST" goal is to provide pointers to parent nodes which the `ProgramTree` does not provide. 
3. The `ProgramTree` has a single type for its nodes, that are split in 4 types in the "AST".

I agree they provide a similar view of the parse-tree, `ProgramTree` could be used if modified to handle point 1. and 2. above (Point 2. would require making `ProgramTree` nodes address stable). Point 3. is a bit more annoying because the type split of the nodes in the "AST" provides a compile time safety when going through the AST to lower.

Reply:
Could semantics be changed to use the "AST" in place of the program tree and save a traversal of the parse tree?